File: src/main/java/com/ruse/world/content/combat/strategy/impl/PlaneFreezer.java
package com.ruse.world.content.combat.strategy.impl;

import com.ruse.engine.task.Task;
import com.ruse.engine.task.TaskManager;
import com.ruse.model.Animation;
import com.ruse.model.Projectile;
import com.ruse.util.Misc;
import com.ruse.world.content.combat.CombatContainer;
import com.ruse.world.content.combat.CombatType;
import com.ruse.world.content.combat.strategy.CombatStrategy;
import com.ruse.model.entity.character.CharacterEntity;
import com.ruse.model.entity.character.npc.NPC;

public class PlaneFreezer implements CombatStrategy {

	@Override
	public boolean canAttack(CharacterEntity entity, CharacterEntity victim) {
		return true;
	}

	@Override
	public CombatContainer attack(CharacterEntity entity, CharacterEntity victim) {
		return null;
	}

	@Override
	public boolean customContainerAttack(CharacterEntity entity, CharacterEntity victim) {
		NPC lakra = (NPC)entity;
		if(victim.getConstitution() <= 0) {
			return true;
		}
		if(lakra.isChargingAttack()) {
			return true;
		}
		lakra.setChargingAttack(true);
		lakra.performAnimation(new Animation((13770)));
		final CombatType attkType = Misc.getRandom(5) <= 2 ? CombatType.RANGED : CombatType.MAGIC;
		lakra.getCombatBuilder().setContainer(new CombatContainer(lakra, victim, 1, 4, attkType, Misc.getRandom(5) <= 1 ? false : true));
		TaskManager.submit(new Task(1, lakra, false) {
			int tick = 0;
			@Override
			public void execute() {
				if(tick == 2) {
					new Projectile(lakra, victim, (attkType == CombatType.RANGED ? 605 : 473), 44, 3, 43, 43, 0).sendProjectile();
					lakra.setChargingAttack(false);
					stop();
				}
				tick++;
			}
		});
		return true;
	}

	@Override
	public int attackDelay(CharacterEntity entity) {
		return entity.getAttackSpeed();
	}

	@Override
	public int attackDistance(CharacterEntity entity) {
		return 5;
	}

	@Override
	public CombatType getCombatType(CharacterEntity entity) {
		return CombatType.MIXED;
	}
}

File: src/main/java/com/realting/world/content/combat/strategy/impl/PlaneFreezer.kt
package com.realting.world.content.combat.strategy.impl

import com.realting.engine.task.Task
import com.realting.engine.task.TaskManager
import com.realting.model.Animation
import com.realting.model.Projectile
import com.realting.model.entity.character.CharacterEntity
import com.realting.model.entity.character.npc.NPC
import com.realting.util.Misc
import com.realting.world.content.combat.CombatContainer
import com.realting.world.content.combat.CombatType
import com.realting.world.content.combat.strategy.CombatStrategy

class PlaneFreezer : CombatStrategy {
    override fun canAttack(entity: CharacterEntity?, victim: CharacterEntity?): Boolean {
        return true
    }

    override fun attack(entity: CharacterEntity?, victim: CharacterEntity?): CombatContainer? {
        return null
    }

    override fun customContainerAttack(entity: CharacterEntity?, victim: CharacterEntity?): Boolean {
        val lakra = entity as NPC?
        if (victim!!.constitution <= 0) {
            return true
        }
        if (lakra!!.isChargingAttack) {
            return true
        }
        lakra.isChargingAttack = true
        lakra.performAnimation(Animation(13770))
        val attkType = if (Misc.getRandom(5) <= 2) CombatType.RANGED else CombatType.MAGIC
        lakra.combatBuilder.container =
            CombatContainer(lakra, victim, 1, 4, attkType, if (Misc.getRandom(5) <= 1) false else true)
        TaskManager.submit(object : Task(1, lakra, false) {
            var tick = 0
            public override fun execute() {
                if (tick == 2) {
                    Projectile(
                        lakra,
                        victim,
                        if (attkType === CombatType.RANGED) 605 else 473,
                        44,
                        3,
                        43,
                        43,
                        0
                    ).sendProjectile()
                    lakra.isChargingAttack = false
                    stop()
                }
                tick++
            }
        })
        return true
    }

    override fun attackDelay(entity: CharacterEntity?): Int {
        return entity!!.attackSpeed
    }

    override fun attackDistance(entity: CharacterEntity): Int {
        return 5
    }

    override fun getCombatType(entity: CharacterEntity): CombatType? {
        return CombatType.MIXED
    }
}